### DISCUSSION TITLE
Allow sizeof on variables

### DISCUSSION THREAD

**Discussion by thatbakamono** [dotnet/csharplang#4429] [Language Ideas]
As described in the title, currently something like this is not possible:
```cs
var numbers = new int[] 
{
	1, 2, 3, 4, 5, 6, 7, 8, 9
};

Console.WriteLine(sizeof(numbers));
```
It's not a big deal with simple arrays because you can solve it like this:
```cs
var numbers = new int[] 
{
	1, 2, 3, 4, 5, 6, 7, 8, 9
};

Console.WriteLine(sizeof(int) * numbers.Length);
```
but it gets a little bit more complicated with stackalloc because you have to store the size somewhere or have magic numbers floating randomly in your code 
```cs
var numbers = stackalloc int[]
{
	1, 2, 3, 4, 5, 6, 7, 8, 9
};
```
It would be very helpful, especially when working with unmanaged code which often expects size.

**Comment by bernd5** [comment]
In the first example, sizeof(members) is not allowed because it is implementation defined. The array does not store only the integers itself, it stores in addition a virtual method table, a length and maybe something more.

But for stackallow, the result is no array. It is a pointer or a Span<T> in new code. You can get the size now via `numbers.Length` ([Link](https://dotnetfiddle.net/eDg61q)).

(Span<T> is a struct which combines a pointer with a length).

**Reply by thatbakamono** [reply]
The proposal isn't about the count of elements of an array (or a stack allocated array), it's about the size that all elements take up in the memory (length * sizeof(T)). I don't think things like length or virtual method table should be counted by the proposed sizeof because they are internals of the language and they are implementation specific.

**Reply by CyrusNajmabadi** [reply]
Can you clarify when/how this is helpful?

**Reply by thatbakamono** [reply]
It's mainly useful for working with unmanaged code, I'm not aware of any other use cases.

**Comment by bernd5** [comment]
With Span of T, you don't need a second variable to store the stack allocated array length. So you can write code similar to heap allocated arrays:

```
using System;

Span<int> numbers = stackalloc int[]
{
	1, 2, 3, 4, 5, 6, 7, 8, 9
};
Console.WriteLine(numbers.Length * sizeof(int));
```

**Reply by thatbakamono** [reply]
The entire point of the proposal is to do ``sizeof(sth)`` instead of ``length * sizeof(T)``

**Reply by bernd5** [reply]
Yes, I know. But in my eyes `sizeof(sth)` could mean different things. It could mean:

- size of the variable itself (for array = sizeof(void*))
- size of user data (like you proposed)
- size of implementation definied memory requirements (including VMT...)
- size after Marshalling

In addition `sizeof` should always return a constant value - which would be not the case with your proposal.

**Reply by thatbakamono** [reply]
Yeah, you are 100% right. It's not constant and a little bit confusing because it could mean 4 different things. Got any idea how it could be solved?

**Reply by theunrepentantgeek** [reply]
> ... it could mean 4 different things. Got any idea how it could be solved?

If there isn't a clearly obvious and correct meaning for **sizeof** then it's probably not a good idea to introduce the feature.

**Comment by x2taptap** [comment]
xDDDDDDDDDDDDD
